### PR TITLE
Rename `rustc_abi::Abi` to `BackendRepr`

### PR DESCRIPTION
**Pull request by workingjubilee**
Remove the confabulation of `rustc_abi::Abi` with what "ABI" actually means by renaming it to `BackendRepr`, and rename `Abi::Aggregate` to `BackendRepr::Memory`. The type never actually represented how things are passed, as that has to have `PassMode` considered, at minimum, but rather it just is how we represented some things to the backend. This conflation arose because LLVM, the primary backend at the time, would lower certain IR forms using certain ABIs. Even that only somewhat was true, as it broke down when one ventured significantly afield of what is described by the System V AMD64 ABI either by using different architectures, ABI-modifying IR annotations, the same architecture **with different ISA extensions enabled**, or other... unexpected delights.

Unfortunately both names are still somewhat of a misnomer right now, as people have written code for years based on this misunderstanding. Still, their original names are even moreso, and for better or worse, this backend code hasn't received as much maintenance as the rest of the compiler, lately. Actually arriving at a correct end-state will simply require us to disentangle a lot of code in order to fix, much of it pointlessly repeated in several places. Thus this is not an "actual fix", just a way to deflect further misunderstandings.

- Acts on long-ago remarks from @eddyb as narrated by @RalfJung in https://github.com/rust-lang/compiler-team/issues/672#issuecomment-1751624436
- Part of https://github.com/rust-lang/rust/issues/119183

r? @jieyouxu

### CODE REVIEW

**Comment by compiler-errors** [comment]
Hate to bikeshed but why is this called "Ir"? We got lots of IRs in the compiler, and the impression I get is that type name is overly vague. It's like naming a type "Kind" with no qualifiers 😃 

LayoutForm? LayoutKind? idk

File: compiler/rustc_abi/src/layout.rs [hunk 1]
```diff
@@ -28,7 +28,7 @@ where
     VariantIdx: Idx,
     F: Deref<Target = &'a LayoutS<FieldIdx, VariantIdx>> + fmt::Debug,
 {
-    let uninhabited = fields.iter().any(|f| f.abi.is_uninhabited());
+    let uninhabited = fields.iter().any(|f| f.ir_form.is_uninhabited());
```
Comment (compiler-errors) [inline compiler/rustc_abi/src/layout.rs:31]:
```suggestion
    let uninhabited = fields.iter().any(|f| f.is_uninhabited());
```

**Review by compiler-errors — COMMENTED** [review]
> At the same time, reduce the need of some backends of rustc to snuffle around the internals of the Layout types by exposing another accessor, LayoutS::is_uninhabited.

Seems like there's quite a few callsites that are still doing `layout.ir_form.is_uninhabited()` rather than going thru the new accessor. I'd also be happy to quickly review that set of changes pulled out of this PR in particular, since it seems somewhat orthogonal to this rename.

Similarly, there are some callsites that call `layout.ir_form.other_method()` for other methods like `is_sized()` that could probably just be turned into direct calls on the layout type?

**Comment by workingjubilee** [comment]
Because that is all it was ever supposed to be.
- It does not actually meaningfully determine layout: there are many scalars that might be represented identically to aggregates and vice versa.
- Likewise, it does not actually determine the calling convention, as there are many IR annotations that might be added via other routes, some of which have much more of an impact.

It primarily was intended to distinguish between "handle as an SSA value (scalar or vector), literally written using the syntactic forms that LLVMIR gives to SSA values, or the IR repr, or interact with it as a pointer to memory".

**Comment by jieyouxu** [comment]
I think I like LayoutForm more than IrForm, for the same reasoning -- we already have *several* IRs in the compiler.

**Comment by compiler-errors** [comment]
> Because that is all it was ever supposed to be.

Cool, but I still think it needs a name that clearly denotes that it is a component of computations that are primarily concerning codegen and layout. Right now that name does not do that at all. "IR" means something completely different to the bulk of the compiler than what you seem to make it mean here.

**Comment by workingjubilee** [comment]
Happy to crank the name generator in my head, just wanted to clarify that going to "LayoutSomething" will add to the confusion.

**Comment by workingjubilee** [comment]
SsaForm or whatever

**Comment by jieyouxu** [comment]
That sounds reasonable to me.

**Comment by workingjubilee** [comment]
Yeah I got the idea of "wait... accessors exist" only midway through this.

**Comment by compiler-errors** [comment]
I'm happy with `SsaForm` or something slightly more descriptive in that direction, but maybe let's wait a few hours (i.e. im gonna go sleep now) to see if anyone else on the backend side has stronger opinions than me.

**Comment by RalfJung** [comment]
`SsaForm` makes little sense to me since a codegen backend doesn't have to use SSA -- for instance, Miri uses this to inform the way it represents local variables, but there's no SSA involved.

In the MCP I suggested names like `StorageClass` or `StorageKind`. We could also use `StorageForm`. To me, "storage" makes sense since it is really a detail about how the rest of the compiler is supposed to store values of this type -- though it is always correct to ignore that and store it in-memory anyway.

**Comment by workingjubilee** [comment]
"Storage" just sounds like a synonym for "Layout" though.

**Comment by workingjubilee** [comment]
The unfortunate problem here is that this type currently _does not answer a question_, but we can't simply discard it without significantly more work.

To call it "Storage" implies it would answer the question of "how will this value be stored?"

But it does not.

**Comment by RalfJung** [comment]
Not at all? "layout" for us has the fairly clear meaning of being about field offsets, and related properties. It's like the layout of a page or website. "storage" has a completely different meaning than that? I don't follow at all TBH.

**Comment by RalfJung** [comment]
The type is a hint, answering the question of how values of this type shall be stored by backends if there are no other constraints that force it to be stored in-memory.

**Comment by workingjubilee** [comment]
> The type is a hint, answering the question of how values of this type shall be stored by backends if there are no other constraints that force it to be stored in-memory.

`PoliteAbiSuggestionIfItIsntTooMuchTrouble`, then???

**Comment by RalfJung** [comment]
Sure. ;)

Maybe `StorageHint`? Though that name would be aspirational, since currently we *do* also use this type to inform the actual call ABI.

**Comment by workingjubilee** [comment]
Hmm. `SomethingHint` does seem the way to go! Currently weighing `StorageHint` versus `AbiHint` (less of a change, annoyingly, but possibly more accurate?) or `ReprHint` (more ambiguous, which could be good?).

**Comment by RalfJung** [comment]
It's currently not just a hint though. For instance it is also the authority on whether this is an unsized type. So I feel like further refactoring would be required before we can use a name like that.

**Comment by bjorn3** [comment]
rustc_abi::Abi does have significant impact on the calling convention, though it indeed doesn't determine the calling convention on it's own. For example the rust abi computation does take rustc_abi::Abi into account when determining the pass mode to use. Also for example PassMode::Direct is only valid when the type is Abi::Scalar or Abi::Vector and similar for PassMode::Pair and Abi::ScalarPair. If you have Abi::Aggregate you have to use PassMode::Cast for passing it in register. I'm not sure what the best name is, but I do completely agree that it should be renamed to something other than rustc_abi::Abi to reduce confusion.

**Comment by workingjubilee** [comment]
My intention was always to make a PR to change the name and later change the code to more accurately reflect the name, yes.

I realize the result will be code that reads incorrectly in a few spots, but that also is intentional, so that the code is provocatively weird, rather than being "confidently wrong". Then, when I fix up the code later and pull a reviewer, rather than needing to explain that the `Abi` type's meaning is bogus, and that I am incrementally changing towards a certain direction, and maybe relitigating that change, the resulting diffs read as obviously reducing the incongruity. Maybe I should also add a `FIXME` to every single field access to `LayoutData` too, I guess.

Obviously if the reviewer is one of the people currently reading this it would go a bit easier,  but I prefer to aim for allowing people that don't have perfect context to review the code.

I don't really want to use something too definite like `Storage`, as I don't know if it will actually describe `Storage`, nor if I _want_ it to, yet. I'm going to be doing a lot to change a lot of adjacent details. So if it takes on a definite form, I'd rather it mostly signal the (original) intention as being about how we represent something to the codegen backend... That suggests another 🎨 choice: `CodegenRepr` maybe.

**Comment by RalfJung** [comment]
I'd prefer "backend" over "codegen" since we have a backend that does not generate code. ;)

**Comment by workingjubilee** [comment]
That would also be fine, I think.

**Comment by compiler-errors** [comment]
So `BackendRepr` it is? :D

**Comment by RalfJung** [comment]
We can always rename again if we find a better name. ;)

**Comment by workingjubilee** [comment]
As I was working through rebasing and updating the commits with the new decision, I kept abbreviating `BackendRepr` to `BackRepr`, so I just shrugged and went with it.

File: compiler/rustc_abi/src/layout/ty.rs [hunk 2]
```diff
@@ -83,8 +83,8 @@ impl<'a> Layout<'a> {
         &self.0.0.variants
     }
 
-    pub fn abi(self) -> Abi {
-        self.0.0.abi
+    pub fn abi(self) -> BackendRepr {
```
Comment (compiler-errors) [inline compiler/rustc_abi/src/layout/ty.rs:86]: naming

File: compiler/rustc_abi/src/lib.rs [hunk 3]
```diff
@@ -1686,7 +1704,7 @@ where
         let LayoutData {
             size,
             align,
-            abi,
+            backend_repr: abi,
```
Comment (compiler-errors) [inline compiler/rustc_abi/src/lib.rs:1707]: rename the `abi` local here?

File: compiler/rustc_codegen_cranelift/src/value_and_place.rs [hunk 4]
```diff
@@ -646,23 +650,28 @@ impl<'tcx> CPlace<'tcx> {
...
+                        BackendRepr::ScalarPair(a_scalar, b_scalar) => {
                             let b_offset =
                                 scalar_pair_calculate_b_offset(fx.tcx, a_scalar, b_scalar);
                             to_ptr.store(fx, val1, flags);
                             to_ptr.offset(fx, b_offset).store(fx, val2, flags);
                         }
-                        _ => bug!("Non ScalarPair abi {:?} for ByValPair CValue", dst_layout.abi),
+                        _ => {
+                            bug!(
+                                "Non ScalarPair abi {:?} for ByValPair CValue",
```
Comment (compiler-errors) [inline compiler/rustc_codegen_cranelift/src/value_and_place.rs:662]:
```suggestion
                                "Non ScalarPair backend repr {:?} for ByValPair CValue",
```

File: compiler/rustc_ty_utils/src/layout.rs [hunk 5]
```diff
@@ -330,17 +333,17 @@ fn layout_of_uncached<'tcx>(
...
+                BackendRepr::Memory { sized: true }
             };
 
             let largest_niche = if count != 0 { element.largest_niche } else { None };
 
             tcx.mk_layout(LayoutData {
                 variants: Variants::Single { index: FIRST_VARIANT },
                 fields: FieldsShape::Array { stride: element.size, count },
-                abi,
+                backend_repr: abi,
```
Comment (compiler-errors) [inline compiler/rustc_ty_utils/src/layout.rs:346]: maybe rename this abi local

**Review by compiler-errors — APPROVED** [review]
Some minor nits, take them or leave them. r=me otherwise.

**Comment by workingjubilee** [comment]
Dropped the random cleanup commit because it has no more path-dependence on this PR (it used to be bigger and involve the Cargo.tomls in the tree).

**Comment by workingjubilee** [comment]
Minor rewording of the initial commit message.

This seems likely to conflict, given it touches most of our subtrees, so

@bors r+ rollup=iffy

**Comment by workingjubilee** [comment]
...
@bors r=compiler-errors

File: tests/ui/abi/c-zst.other-linux.stderr
Comment (workingjubilee) [inline tests/ui/abi/c-zst.other-linux.stderr:1]: the other variants of this went unblessed.

**Comment by workingjubilee** [comment]
Rewrote the platform-specific ZST ABI tests to now do cross-compilation, as they each only really need a single Sized lang-item. They are now *also* blessed.

File: tests/ui/abi/c-zst.rs [hunk 6]
```diff
@@ -1,27 +1,40 @@
-//@ revisions: other other-linux x86_64-pc-windows-gnu s390x-linux sparc64-linux powerpc-linux
 //@ normalize-stderr-test: "(abi|pref|unadjusted_abi_align): Align\([1-8] bytes\)" -> "$1: $$SOME_ALIGN"
+
+// Ignore the ZST revisions
```
Comment (RalfJung) [inline tests/ui/abi/c-zst.rs:3]: I don't understand the comment. What are we supposed to ignore?

Comment (RalfJung) [reply tests/ui/abi/c-zst.rs:3]: The old code had two sections,"ignore the ZST" and "pass the ZST indirectly". That structure seems to have been lost and now it's just confusing.

Comment (workingjubilee) [reply tests/ui/abi/c-zst.rs:3]: I found the previous structure confusing too, tbh.

Comment (workingjubilee) [reply tests/ui/abi/c-zst.rs:3]: @RalfJung I don't feel confident that our handling of passing the ZST by indirection is correct for the ABIs in question.

Comment (RalfJung) [reply tests/ui/abi/c-zst.rs:3]: > I found the previous structure confusing too, tbh.

There are two kinds of targets: those where ZST are ignored, and those where they are passed indirectly (which may seem nonsensical but it's what it is). It is helpful to group the revisions for the test by which category a target falls into.

Comment (workingjubilee) [reply tests/ui/abi/c-zst.rs]: @RalfJung I did a bit more studying to satisfy my curiosity a bit more as to exactly _why_ so many ABIs decide ZSTs consume space, and organized the revisions accordingly again, plus added some remarks as to the history.

Comment (RalfJung) [reply tests/ui/abi/c-zst.rs]: Thanks!

I was going to suggest a better place for the comment would be the code in rustc that implements this ABI logic. That used to be one central place but https://github.com/rust-lang/rust/commit/b1493ba5194fcc5cfe4f6315db288e4e18509110 recently moved it to the corresponding targets... which is probably a good call but makes it harder to to put a good comment in a central place.

Comment (workingjubilee) [reply tests/ui/abi/c-zst.rs]: Yeah, "the code may or may not be in one place" is basically the reason I've been leaving these sorts of comments by the test cases. Tests in many ways describe the spec that we're adhering to, but the implementation can be whatever... including "spread out across 100 modules".

I may find a better home for some of this text at some point, but not today.

**Comment by workingjubilee** [comment]
post-hoc note: If this commit broke your custom driver, codegen backend, or other compiler extension, try to move touching on the `layout.backend_repr` to the convenience functions now available, like `LayoutData::is_uninhabited`.

**Pull request merged**